### ISSUE TITLE
Recycled buffers in XMLStreamWriter are colliding when multiple when used in an Akka stream

### ISSUE THREAD

**Issue by jphelp32** [FasterXML/woodstox#82]
I'll start w/the caveat that I have not been able to produce a unit test case to illustrate this issue. But what I'm observing in our application is that the recycled char buffers (pretty sure it's the small copy buffers, but could potentially be the others), are being concurrently used by multiple threads in an Akka stream, and the result is corrupted output. In particular in the namespace writing, in my application.
I've so far been unable to reproduce the problem using aalto, but I'm not certain it also isn't affected given that it also uses the ThreadLocal recycled buffers.
Hoping you might have an inkling as to how to attempt to reproduce outside of Akka and my application? Or perhaps this is already a known issue that woodstox should not be used under fork-join parallelism?  (However, I did try changing my akka configuration to use a fixed threadpool dispatcher and the problem still occurs.)
I've locally recompiled woodstox w/some debug logging added and I have some log files that demonstrate the problem. Don't want to presume upon you to review them, but I mention in case it might help.

**Comment by jphelp32** [comment]
Currently using woodstox-core:5.3.0, but also tested 5.2.1 and 6.0.0 and they are also affected.

**Comment by cowtowncoder** [comment]
Ok. Unfortunate timing, on one hand (since I just pushed 6.0.0 recently), but possibly fortunate since I think I went through similar problem with Jackson.

So: one assumption with current buffer recycling is that parser/generator instances are only ever used from a single thread: not just at any given time (i.e. single-threaded access), but during life-cycle of parser/generator instance. If so, `ThreadLocal` suffices.
In case of Jackson this was generally true, except that when async-parsing was implemented, usage pattern changed, and problem was found. Solution was to use atomic references in addition to `ThreadLocal`/`SoftReference`, and I think that should solve the issue (full fix will be in 2.10.0).

I was actually wondering if I should port changes to Woodstox' recycler (which is where Jackson originally got it impl from), but since I had not heard of any issues I hadn't.

But sounds like there is now use case.

Now: do you think it would be easy for you to test a fix, if I come up with what I think should work? Does not have to be automated.

**Comment by jphelp32** [comment]
Yeah, I'm able to consistently reproduce within my application using an aggressive gatling script.
I'm curious though..  with aalto, I cannot reproduce with that same script. And that is using the same ThreadLocal/SoftReference strategy as far as I can tell. Any thoughts as to why that might be?

**Comment by cowtowncoder** [comment]
Hmmh. That is interesting -- code in `BufferRecycler` looks identical, so I would expect behavior to be similar. I assume access pattern for either obtaining or releasing might differ.

**Comment by cowtowncoder** [comment]
One thing wrt Aalto being less prone could be that it does fewer copies in general, and bundles local-name/namespace-URI combination so it could just have less time to have collision or something. So perhaps it is just significantly less common to get corruption, but not immune.

**Comment by jphelp32** [comment]
That's sorta what I thought might be the case. And why I didn't just switch over to aalto and move on.

**Comment by cowtowncoder** [comment]
:)

It'd be good to resolve the problem, still.

One more question: how is the shared access from multiple threads done? As in, are parsers/generators (problem is on generator side, right?) being shared or passed across processing threads? Since both operations are sequential, I assume there is some hand-over or. ..?

**Comment by jphelp32** [comment]
Problem is on the generator side, yes. Handoff is happening entirely by the akka stream/actors framework. I'm not exactly sure under the covers how it assigns threads to actors, etc.
I tried to write a unit test to reproduce, taking akka out of the equation, but could not reproduce the problem. Though I'm not very familiar with java concurrency and ExecutorService etc, to know whether or not my test was a good representation of what akka is doing.

**Comment by jphelp32** [comment]
To answer your earlier question, it appears to be the medium cbuffer that I'm having issues with.

**Comment by cowtowncoder** [comment]
Was included in `6.0.1` release
